internal/gatewayapi/securitypolicy.go

```diff
@@ -0,0 +1,250 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package gatewayapi
+
+import (
+	"fmt"
+	"sort"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	gwv1a2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+	gwv1b1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	egv1a1 "github.com/envoyproxy/gateway/api/v1alpha1"
+	"github.com/envoyproxy/gateway/internal/status"
+	"github.com/envoyproxy/gateway/internal/utils/ptr"
+)
+
+func securityPolicyHasSectionName(policy *egv1a1.SecurityPolicy) bool {
+	return policy.Spec.TargetRef.SectionName != nil
+}
+
+func ProcessSecurityPolicies(securityPolicies []*egv1a1.SecurityPolicy,
```

Comment:
need to be able to support 3 levels of targetRefs
* Gateway
* Gateway with sectionName (Listener)
* HTTPRoute or GRPCRoute